routemaster/cron.py

```diff
@@ -19,35 +19,79 @@
 )
 from routemaster.state_machine import process_action_retries
 
+IsExitingCheck = Callable[[], bool]
+CronProcessor = Callable[
+    [App, State, StateMachine, IsExitingCheck],
+    None,
+]
+
+logger = logging.getLogger(__name__)
+
 
-def _retry_action(app, state, state_machine, process_part) -> None:
+def _retry_action(app, state, state_machine, should_terminate) -> None:
     print(f"Processing action retries for {state.name}")
-    process_action_retries(app, state, state_machine, process_part)
+    process_action_retries(app, state, state_machine, should_terminate)
 
 
-def _trigger_gate(app, state, state_machine, process_part) -> None:
+def _trigger_gate(app, state, state_machine, should_terminate) -> None:
     print(f"Processing interval/time trigger for {state.name}")
 
 
-def _retry_metadata_updates(app, state, state_machine, process_part) -> None:
+def _retry_metadata_updates(
+    app,
+    state,
+    state_machine,
+    should_terminate,
+) -> None:
     print(f"Processing metadata update trigger for {state.name}")
 
 
+class _Process(NamedTuple):
+    fn: CronProcessor
+    state: State
+    app: App
+    state_machine: StateMachine
+    is_terminating: IsExitingCheck
+
+    def __call__(self):
+        logger.info(
+            f"Started cron {self.fn.__name__} for state {self.state.name} in "
+            f"{self.state_machine.name}",
+        )
+
+        try:
+            time_start = time.time()
+            self.fn(
+                self.app,
+                self.state,
+                self.state_machine,
+                self.is_terminating,
+            )
+            duration = time.time() - time_start
+        except Exception as e:
+            print(e)
```

Comment:
Prefer `logger.exception()` (which also means we can drop the `as e` on the previous line).